File: WdtcCore/src/main/java/org/wdt/wdtc/core/game/config/VersionInfo.java
package org.wdt.wdtc.core.game.config;

import com.google.gson.annotations.SerializedName;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;
import org.wdt.wdtc.core.download.fabric.FabricDonwloadInfo;
import org.wdt.wdtc.core.download.forge.ForgeDownloadInfo;
import org.wdt.wdtc.core.download.infterface.ModDownloadInfoInterface;
import org.wdt.wdtc.core.download.quilt.QuiltInstallTask;
import org.wdt.wdtc.core.game.Launcher;
import org.wdt.wdtc.core.utils.ModUtils;

@Setter
@Getter
@ToString
public class VersionInfo {
  @SerializedName("GameVersionNumber")
  private String VersionNumber;
  @SerializedName("ModKind")
  private ModUtils.KindOfMod kind;
  @SerializedName("ModVersionNumber")
  private String ModVersion;

  public VersionInfo(Launcher launcher) {
    this.VersionNumber = launcher.getVersionNumber();
    this.kind = launcher.getKind();
    ModDownloadInfoInterface info = ModUtils.getModDownloadInfo(launcher);
    if (info != null) {
      this.ModVersion = info.getModVersion();
    } else {
      this.ModVersion = null;
    }
  }


  public Launcher getLauncher() {
    Launcher launcher = new Launcher(getVersionNumber());
    switch (getKind()) {
      case FORGE -> launcher.setForgeModDownloadInfo(new ForgeDownloadInfo(launcher, getModVersion()));
      case FABRIC -> launcher.setFabricModInstallInfo(new FabricDonwloadInfo(launcher, getVersionNumber()));
      case QUILT -> launcher.setQuiltModDownloadInfo(new QuiltInstallTask(launcher, getVersionNumber()));
    }
    return launcher;
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/game/config/VersionInfo.kt
package org.wdt.wdtc.core.game.config

import org.wdt.wdtc.core.download.fabric.FabricDonwloadInfo
import org.wdt.wdtc.core.download.forge.ForgeDownloadInfo
import org.wdt.wdtc.core.download.game.VersionNotFoundException
import org.wdt.wdtc.core.download.quilt.QuiltInstallTask
import org.wdt.wdtc.core.game.Version
import org.wdt.wdtc.core.utils.KindOfMod
import org.wdt.wdtc.core.utils.modDownloadInfo

data class VersionInfo(
  val versionNumber: String,

  val kind: KindOfMod,

  var modVersion: String? = null
) {

  constructor(version: Version) : this(
    version.versionNumber,
    version.kind,
    version.modDownloadInfo?.modVersion
  )

  val version: Version
    get() {
      val modVersion = this.modVersion ?: throw NullPointerException("$modVersion is null")
      val version = Version(versionNumber).apply {
        when (kind) {
          KindOfMod.FORGE -> forgeModDownloadInfo = ForgeDownloadInfo(this, modVersion)
          KindOfMod.FABRIC -> fabricModInstallInfo = FabricDonwloadInfo(this, modVersion)
          KindOfMod.QUILT -> quiltModDownloadInfo = QuiltInstallTask(this, modVersion)
          else -> {
            throw VersionNotFoundException("Game mod not found")
          }
        }
      }
      return version
    }
}

val Version.versionInfo
  get() = VersionInfo(this)
